File: doric-flutter/doric/lib/doric/engine/doric_native_driver.dart
import '../../flutter_jscore.dart';
import '../doric_registry.dart';
import 'doric_constant.dart';
import 'doric_engine.dart';
import 'doric_driver.dart';

class DoricNativeDriver with IDoricDriver {
  DoricNativeDriver._() {
    jsEngine = DoricJSEngine();
  }

  static final _instance = DoricNativeDriver._();

  factory DoricNativeDriver.getInstance() => _instance;

  DoricJSEngine jsEngine;

  @override
  bool createContext(String contextId, String script, String source) {
    try {
      jsEngine.prepareContext(contextId, script, source);
      return true;
    } catch (e) {
      return false;
    }
  }

  @override
  bool destroyContext(String contextId) {
    try {
      jsEngine.destroyContext(contextId);
      return true;
    } catch (e) {
      return false;
    }
  }

  getJSContext() {
    return jsEngine.getJSContext();
  }

  DoricRegistry getRegistry() {
    return jsEngine.getRegistry();
  }

  @override
  JSValue invokeContextEntityMethod(
      String contextId, String method, List<JSValue> args) {
    var theArgs = [
      JSValue.makeString(jsEngine.getJSContext(), contextId),
      JSValue.makeString(jsEngine.getJSContext(), method)
    ];
    theArgs.addAll(args);
    return invokeDoricMethod(DoricConstant.DORIC_CONTEXT_INVOKE, theArgs);
  }

  @override
  JSValue invokeDoricMethod(String method, List<JSValue> args) {
    return jsEngine.invokeDoricMethod(method, args);
  }
}

File: doric-Qt/doric/DoricNativeDriver.cpp
#include <functional>

#include "DoricNativeDriver.h"
#include "async/DoricAsyncCall.h"
#include "utils/DoricConstant.h"

void DoricNativeDriver::invokeContextEntityMethod(QString contextId,
                                                  QString method,
                                                  QVariantList args) {
  args.insert(0, QVariant(contextId));
  args.insert(1, QVariant(method));
  invokeDoricMethod(DoricConstant::DORIC_CONTEXT_INVOKE, args);
}

void DoricNativeDriver::invokeDoricMethod(QString method, QVariantList args) {
  return DoricAsyncCall::ensureRunInThreadPool(
      &jsEngine.mJSThreadPool, [this, method, args] {
        this->jsEngine.invokeDoricMethod(method, args);
      });
}

DoricAsyncResult *DoricNativeDriver::asyncCall(std::function<void()> lambda,
                                               DoricThreadMode mode) {
  switch (mode) {
  case UI:
    DoricAsyncCall::ensureRunInMain(lambda);
    break;
  case JS:
    DoricAsyncCall::ensureRunInThreadPool(&jsEngine.mJSThreadPool, lambda);
    break;
  }
  return NULL;
}

void DoricNativeDriver::createContext(QString contextId, QString script,
                                      QString source) {
  DoricAsyncCall::ensureRunInThreadPool(
      &jsEngine.mJSThreadPool, [this, contextId, script, source] {
        this->jsEngine.prepareContext(contextId, script, source);
      });
}

void DoricNativeDriver::destroyContext(QString contextId) {}

DoricRegistry *DoricNativeDriver::getRegistry() {
  return this->jsEngine.getRegistry();
}
